### PR TITLE
Azure communication service build

### PR DESCRIPTION
**Pull request by likelymichael**
Tested in personal org deploying to 3 different `data_locations` and tearing them down as well.

### CODE REVIEW

File: src/main.tf [hunk 1]
```diff
@@ -0,0 +1,33 @@
+locals {
+  geo_location_map = {
```
Comment (WillBeebe) [inline src/main.tf:6]: Can we put this map in the `massdriver.yaml` as a type string with the `oneOf` pattern we used yesterday with the `taint`?

File: src/main.tf [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    "India"         = "centralindia"
+    "Japan"         = "japaneast"
+    "Korea"         = "koreacentral"
+    "Norway"        = "norwayeast"
+    "Switzerland"   = "switzerlandnorth"
+    "UAE"           = "uaenorth"
+    "UK"            = "uksouth"
+    "United States" = "centralus"
+  }
+  location = lookup(local.geo_location_map, var.location, "")
```
Comment (WillBeebe) [inline src/main.tf:24]: A user can add something to the `massdriver.yaml` and forget to add it to the list above. In that case validation will pass but `location` will be `""` and provisioning will fail.

Before I really learned JSON schema, I think I did this in a few places in GCP as well. Mostly for enums in static lists like availability `Highly Available => HIGHLY_AVAILABLE`. I think for Firestore, I do this with locations and if that yaml changes it would be the same issue. I'll plop a ticket in the backlog for me to check / update Firestore.

File: src/main.tf [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+resource "azurerm_resource_group" "main" {
+  name     = var.md_metadata.name_prefix
+  location = local.location
+}
+
+resource "azurerm_communication_service" "main" {
+  name                = var.md_metadata.name_prefix
+  resource_group_name = azurerm_resource_group.main.name
+  data_location       = var.location
```
Comment (WillBeebe) [inline src/main.tf:36]: Should this be?
```suggestion
  data_location       = local.location
```

File: src/_artifacts.tf [hunk 4]
```diff
@@ -0,0 +1,26 @@
+resource "massdriver_artifact" "api" {
+  field                = "api"
+  provider_resource_id = azurerm_communication_service.main.id
+  name                 = "Azure Communication Services ${azurerm_communication_service.main.name} (${azurerm_communication_service.main.id})"
```
Comment (WillBeebe) [inline src/_artifacts.tf]: Not just Azure! The `name`s everywhere are long and kinda gross. Users can change the name of an artifact _after_ it's been created, and further provisioning runs will _not_ update the name. So the first run is our best shot at something human-readable before they change it to "server number 5" or whatever they'd like.

TL;DR this is what this pattern looks like when rendered.

```
Virtual Network beebeaz-dev-central-b762 (/subscriptions/4718857a-5dbe-452c-a172-21a3ae74304f/resourceGroups/beebeaz-dev-central-b762/providers/Microsoft.Network/virtualNetworks/beebeaz-dev-central-b762)
```

vs

```
Virtual Network beebeaz-dev-central-b762 
```

The resource IDs in any cloud are v v long and they're passed as the `provider_resource_id` anyway. Personally, I think it looks better w/o it in the `name`.

```suggestion
  name                 = "Azure Communication Services ${azurerm_communication_service.main.name}"
```

File: massdriver.yaml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  examples: []
+  required:
+    - location
+  properties:
+    location:
+      type: string
+      title: Data location
+      description: Specify a geography to retain chat messages and resource data.
+      default: United States
+      enum:
```
Comment (WillBeebe) [inline massdriver.yaml:18]: see note below about `oneOf` and `title` / `const`

File: src/_artifacts.tf [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  artifact = jsonencode(
+    {
+      data = {
```
Comment (WillBeebe) [inline src/_artifacts.tf:7]: Maybe it's another pass, but how do we authenticate w/ this service?

**Review by WillBeebe — COMMENTED** [review]
Looking good, couple of comments.

File: src/main.tf (hunk 1 above)
Comment (likelymichael) [reply src/main.tf:6]: Good idea, I'll work on this change now

Comment (likelymichael) [reply src/main.tf:6]: Actually, now that I looked at it again, I don't think switching it to a `oneOf` would work because what the map is doing is setting an actual Azure region (like `uksouth`) to the resource group for deployment based on the `data_location` the customer chooses for the Azure communication service.

Setting it up as the `title/const` wouldn't solve that problem.

File: src/main.tf (hunk 2 above)
Comment (likelymichael) [reply src/main.tf:24]: Would the best approach be to remove the `""` from L20?

File: src/main.tf (hunk 3 above)
Comment (likelymichael) [reply src/main.tf:36]: No, because we want the data location of the Comms service to be one of these:
image
Ref: https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/communication_service

And then we use the map to convert their `data_location` to an actual Azure region for the resource group deployment.

File: src/_artifacts.tf (hunk 4 above)
Comment (likelymichael) [reply src/_artifacts.tf]: Good point! I'm good with this change.

File: src/_artifacts.tf (hunk 6 above)
Comment (likelymichael) [reply src/_artifacts.tf:7]: EDIT: I only mentioned how to connect to it, which doesn't answer your question. To authenticate, they can use a service principal, SSO, or log-in via web browser: https://learn.microsoft.com/en-us/azure/communication-services/quickstarts/identity/service-principal-from-cli#authenticate-a-registered-application-in-the-development-environment

The `endpoint` (or `hostname` in the `artifacts.tf` file) is set in their function or code, something like this:
```Javascript
const { DefaultAzureCredential } = require("@azure/identity");
const { CommunicationIdentityClient, CommunicationUserToken } = require("@azure/communication-identity");
const { SmsClient, SmsSendRequest } = require("@azure/communication-sms");

const credential = new DefaultAzureCredential();

async function createIdentityAndIssueToken(resourceEndpoint) {
    const client = new CommunicationIdentityClient(resourceEndpoint, credential);
    return await client.createUserAndToken(["chat"]);
}

async function sendSms(resourceEndpoint, fromNumber, toNumber, message) {
    const smsClient = new SmsClient(resourceEndpoint, credential);
    const sendRequest = {
        from: fromNumber,
        to: [toNumber],
        message: message
    };
    return await smsClient.send(
        sendRequest,
        {} //Optional SendOptions
    );
}

async function main() {
    // You can find your endpoint and access key from your resource in the Azure portal
    // e.g. "https://<RESOURCE_NAME>.communication.azure.com";
    const endpoint = "https://<RESOURCE_NAME>.communication.azure.com/"

    
    console.log("Retrieving new Access Token, using Service Principals");
    const result = await createIdentityAndIssueToken(endpoint);
    console.log(`Retrieved Access Token: ${result.token}`);

    console.log("Sending SMS using Service Principals");

    // You will need a phone number from your resource to send an SMS.
    const smsResult = await sendSms(endpoint, "<FROM NUMBER>", "<TO NUMBER>", "Hello from Service Principals");
    console.log(`SMS ID: ${smsResult[0].messageId}`);
    console.log(`Send Result Successful: ${smsResult[0].successful}`);
}

main();
```

Comment (WillBeebe) [reply src/_artifacts.tf:7]: I'll read more up on that, and this looks like a sweet future app example ^.

My thoughts here are, if I use a service principal, which one? How do I give that one the right permissions to access this service? If the other flow is they go to the endpoint and then log in with their username and password, where does their user get permission for this service?

Comment (likelymichael) [reply src/_artifacts.tf:7]: All very good questions. Let me look into this and get back to you. In the meantime, would you mind reviewing the other comments I responded to?

File: src/main.tf (hunk 3 above)
Comment (WillBeebe) [reply src/main.tf:36]: V interesting. Had a chat w/ Michael and the `local.location` is needed for the `resource_group`. We create a resource group that's semi-close and as-good-as-possible to the data location of the Comms service. 👍

File: src/main.tf (hunk 2 above)
Comment (WillBeebe) [reply src/main.tf:24]: It sounds like these all map to a specific thing, so there shouldn't be a missing one. Only in the case that a new data location is added and we don't pick a region for the resource group.

💡 instead of `""` we could have that be a default location, but that might be a future landmine. Like, can I have a resource group in Asia but a Comm Service in the `UK`?

Comment (likelymichael) [reply src/main.tf:24]: I will test this potential outcome and let you know

File: src/_artifacts.tf (hunk 6 above)
Comment (likelymichael) [reply src/_artifacts.tf:7]: Ok so here's the [breakdown](https://learn.microsoft.com/en-us/azure/communication-services/concepts/authentication#authentication-options) from what I've been able to find so far.

I've been able to auth by creating an identity w/ Azure AD and then giving it a token. This can be created within the SDK by setting the endpoint as a `const` and also setting [a few environment variables](https://learn.microsoft.com/en-us/azure/communication-services/quickstarts/identity/service-principal-from-cli#set-environment-variables) for an Azure Service Principal using `AZURE_CLIENT_ID`, `AZURE_CLIENT_SECRET`, and `AZURE_TENANT_ID`. I've tested using this method to auth, and it works. I noticed there was another option for an Azure account (human) using `AZURE_USERNAME` and `AZURE_PASSWORD`, but I haven't been able to get that to work (I think because I have 2FA enabled on my account). 

EDIT: As far as permissions go, the service principal used within environment variables needs to be a `Contributor` scoped to the subscription.

File: src/main.tf (hunk 2 above)
Comment (likelymichael) [reply src/main.tf:24]: Yes, it is possible for the Azure Communication Service to deploy to a geography like UK and for the resource group it's attached to be located in somewhere like Japan East.

File: src/_artifacts.tf (hunk 6 above)
Comment (coryodaniel) [reply src/_artifacts.tf:7]: Do we need to add a 'policies' block to API artdef to support authorization?

File: src/main.tf (hunk 2 above)
Comment (coryodaniel) [reply src/main.tf:24]: if its blank because someone entered something weird, we should let it crash... they should get a validation error from the API, so highly unlikely

File: src/_artifacts.tf (hunk 6 above)
Comment (coryodaniel) [reply src/_artifacts.tf:7]: The above info and link would be great in the operator guide.

@chrisghill @WillBeebe @xpositivityx what do you guys think about adding  `operator.draft.md` that we can keep note sin until a guide is ready to be published.

Comment (xpositivityx) [reply src/_artifacts.tf:7]: I love this idea.

Comment (likelymichael) [reply src/_artifacts.tf:7]: That depends on what developing an app with the azure-communication-service will look like in Massdriver. From the [docs](https://learn.microsoft.com/en-us/azure/communication-services/quickstarts/identity/service-principal-from-cli), they state that service principal credentials will need to be set as env vars, and also that service principal needs the `Contributor` role. From a developer's perspective, what would it look like to them? From my end, I set env vars in my terminal and used the command `node index.js` [to test](https://learn.microsoft.com/en-us/azure/communication-services/quickstarts/identity/service-principal?pivots=programming-language-javascript#run-the-program).

I was thinking the customer would either use the Massdriver service principal, which already has Owner permissions, to auth, _or_ we can create a service principal with `Contributor` for them and output the info they need. The `Contributor` role seems like it needs to be scoped to the subscription from what I can tell in the docs.

Long story short, I don't think the RBAC in the artifact file is needed.

File: src/main.tf (hunk 2 above)
Comment (likelymichael) [reply src/main.tf:24]: Ok, I'll leave as is then.

File: src/_artifacts.tf (hunk 6 above)
Comment (WillBeebe) [reply src/_artifacts.tf:7]: > Do we need to add a 'policies' block to API artdef to support authorization?

I think that makes sense here. The docs for Com Service say to use the `DefaultCredential()` method in the SDK. Like @mclacore mentions, it needs Contributor.  

```json
{
  "data": {
    "type": "object",
    "required": [
      "api"
    ],
    "properties": {
      "etc": {},
      "api": {},
      "security": {
        "title": "Security details",
        "type": "object",
        "anyOf": [
          {
            "$ref": "../types/aws-security.json"
          },
          {
            "$ref": "../types/azure-security.json"
          },
          {
            "$ref": "../types/gcp-security.json"
          }
        ]
      }
    }
  }
}
```

It would be nice is there was _any role_ that was more restrictive than the Contributor role. 

```hcl
  data_security = {
    iam = {
      "read/write" = {
        role  = "Contributor"
        scope = azurerm_communication_service.main.id
      },
    }
  }
```

Comment (WillBeebe) [reply src/_artifacts.tf:7]: If that's too much cloud for the api artifact definition we can make an artifact definition just for the azure com service.

Comment (likelymichael) [reply src/_artifacts.tf:7]: I'm pretty sure that would need to be scoped to the subscription: https://learn.microsoft.com/en-us/azure/communication-services/quickstarts/identity/service-principal-from-cli#creating-an-azure-active-directory-registered-application

So it would probably look like:
```
  data_security = {
    iam = {
      "read/write" = {
        role  = "Contributor"
        scope = "/subscriptions/${var.azure_service_principal.data.subscription_id}"
      },
    }
  }
```

Comment (coryodaniel) [reply src/_artifacts.tf:7]: We definitely don't want to run user apps using the SP they give us during onboarding. Apps should be tightly scoped using Auto IAM so they have principle of least priv.

I wouldn't worry about human user auth, this is something we have dismissed supporting today. We have a grander scope to add support for that later this year after RBAC.

They way this _should_ work is:

1. the bundle provisions the resource and outputs the IAM policies/scopes/etc (subscription Contributor scope?)
2. mdxc creates an identity (role, principal, whatev) for the app and selects the policy/scope to bind.

Comment (likelymichael) [reply src/_artifacts.tf:7]: Ok cool. I'm understanding now. Based on the [docs](https://learn.microsoft.com/en-us/azure/communication-services/quickstarts/identity/service-principal-from-cli#creating-an-azure-active-directory-registered-application), we can add a policies block based on Beebe's example.

Back to your original question:
> Do we need to add a 'policies' block to API artdef to support authorization?

I think we do. What would be the best approach for that?

Comment (likelymichael) [reply src/_artifacts.tf:7]: I tested auth to comm service SDK using a service principal with `Contributor` role to the comm service itself only, not to the subscription, and I was issued a token. So I think scoping to the comm service only is the best way, even though the docs scope to the subscription.

Comment (chrisghill) [reply src/_artifacts.tf:7]: > This can be created within the SDK by setting the endpoint as a const and also setting [a few environment variables](https://learn.microsoft.com/en-us/azure/communication-services/quickstarts/identity/service-principal-from-cli#set-environment-variables) for an Azure Service Principal using `AZURE_CLIENT_ID`, `AZURE_CLIENT_SECRET`, and `AZURE_TENANT_ID`. I've tested using this method to auth, and it works. I noticed there was another option for an Azure account (human) using `AZURE_USERNAME` and `AZURE_PASSWORD`

I'm curious if this works for managed identities (system assigned and user assigned). It seems we are trying to move that direction since its considered "best-practice" by Azure and is credential-less.

Comment (likelymichael) [reply src/_artifacts.tf:7]: Identity SDK doc for ref: https://azuresdkdocs.blob.core.windows.net/$web/javascript/azure-communication-identity/1.0.0/index.html

Comment (coryodaniel) [reply src/_artifacts.tf:7]: So what would be the `role` and `scope` we would be emitting in the security block of the API artifact? role is `Contributor`? Not sure I follow the scope... is that the subscription?

Comment (likelymichael) [reply src/_artifacts.tf:7]: The scope will be to the `azurerm_communication_service.main.id`. I tested it locally on my end, and I was able to auth and generate a token with a service principal scoped to the communication service only.

**Comment by likelymichael** [comment]
Pertinent info to auth a user-assigned identity to Azure Communication Service from within an app: https://github.com/Azure/azure-sdk-for-js/blob/@azure/identity_3.1.3/sdk/identity/identity/samples/AzureIdentityExamples.md#authenticating-a-user-assigned-managed-identity-with-defaultazurecredential

File: src/main.tf [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    "Norway"        = "norwayeast"
+    "Switzerland"   = "switzerlandnorth"
+    "UAE"           = "uaenorth"
+    "UK"            = "uksouth"
+    "United States" = "westus2"
+  }
+  location = lookup(local.geo_location_map, var.location, "")
+}
+
+resource "azurerm_resource_group" "main" {
```
[the remark this answers is not in the document]
Comment (likelymichael) [reply src/main.tf:27]: Added tags

File: src/_artifacts.tf [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        infrastructure = {
+          ari = azurerm_communication_service.main.id
+        }
+        authentication = {
+          endpoint = "https://${azurerm_communication_service.main.name}.communication.azure.com"
+        }
+        security = {
+          iam = {
+            "read_write" = {
```
Comment (WillBeebe) [inline src/_artifacts.tf:16]: nice 👌

**Review by WillBeebe — APPROVED** [review]
PR is good to merge AFAIC. We might follow with small edits but overall LGTM.

Blocked by:
https://github.com/massdriver-cloud/artifact-definitions/pull/144
